File: java/lab5/src/productcreation/ProductBuilder.java
package src.productcreation;

import java.time.LocalDateTime;

import src.baseabstractions.Builder;
import src.baseobjects.Coordinates;
import src.baseobjects.Person;
import src.baseobjects.Product;
import src.baseobjects.UnitOfMeasure;

/**
 * Builder class for constructing Product objects with a fluent interface.
 *
 * @param <T> the type of Product object to be built, must extend Product
 */
public class ProductBuilder<T extends Product> implements Builder<T> {
    private T product;

    /**
     * Constructs a ProductBuilder initialized with the specified product.
     *
     * @param product the product object to be used as the base for building
     */
    public ProductBuilder(T product) {
        this.product = product;
    }

    /**
     * Sets the ID of the product.
     *
     * @param id the ID to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setId(long id) {
        product.setId(id);
        return this;
    }

    /**
     * Sets the name of the product.
     *
     * @param name the name to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setName(String name) {
        product.setName(name);
        return this;
    }

    /**
     * Sets the coordinates of the product.
     *
     * @param coordinates the coordinates to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setCoordinates(Coordinates coordinates) {
        product.setCoordinates(coordinates);
        return this;
    }

    /**
     * Sets the creation date of the product.
     *
     * @param creationDate the creation date to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setCreationDate(LocalDateTime creationDate) {
        product.setCreationDate(creationDate);
        return this;
    }

    /**
     * Sets the price of the product.
     *
     * @param price the price to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setPrice(long price) {
        product.setPrice(price);
        return this;
    }

    /**
     * Sets the part number of the product.
     *
     * @param partNumber the part number to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setPartNumber(String partNumber) {
        product.setPartNumber(partNumber);
        return this;
    }

    /**
     * Sets the manufacture cost of the product.
     *
     * @param manufactureCost the manufacture cost to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setManufactureCost(Long manufactureCost) {
        product.setManufactureCost(manufactureCost);
        return this;
    }

    /**
     * Sets the unit of measure of the product.
     *
     * @param unitOfMeasure the unit of measure to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setUnitOfMeasure(UnitOfMeasure unitOfMeasure) {
        product.setUnitOfMeasure(unitOfMeasure);
        return this;
    }

    /**
     * Sets the owner of the product.
     *
     * @param owner the owner to set
     * @return this builder instance for method chaining
     */
    public ProductBuilder<T> setOwner(Person owner) {
        product.setOwner(owner);
        return this;
    }

    /**
     * Resets the builder to its initial state.
     * Currently does nothing as the builder uses an existing product reference.
     */
    public void reset() {

    }

    /**
     * Returns the constructed product object.
     *
     * @return the built product of type T
     */
    public T create() {
        return product;
    }

}

File: java/lab6/client/src/main/kotlin/objectCreation/ProductBuilder.kt
package objectCreation

import abstractions.Builder
import objects.Coordinates
import objects.Person
import objects.Product
import objects.UnitOfMeasure

/**
 * Builds a [Product] instance field by field.
 * The [id] and [creationDate] fields are not set here — they are assigned by the server.
 */
class ProductBuilder : Builder<Product> {

    private var name: String = ""
    private var coordinates: Coordinates = Coordinates(0, 0.0)
    private var price: Long = 1
    private var partNumber: String? = null
    private var manufactureCost: Long? = null
    private var unitOfMeasure: UnitOfMeasure? = null
    private var owner: Person = Person("placeholder", 1, objects.Color.GREEN)

    override fun reset() {
        name = ""
        coordinates = Coordinates(0, 0.0)
        price = 1
        partNumber = null
        manufactureCost = null
        unitOfMeasure = null
        owner = Person("placeholder", 1, objects.Color.GREEN)
    }

    /**
     * Sets the product [name].
     *
     * @param name non-blank product name
     * @return this builder
     */
    fun setName(name: String): ProductBuilder {
        this.name = name
        return this
    }

    /**
     * Sets the product [coordinates].
     *
     * @param coordinates non-null coordinates
     * @return this builder
     */
    fun setCoordinates(coordinates: Coordinates): ProductBuilder {
        this.coordinates = coordinates
        return this
    }

    /**
     * Sets the product [price].
     *
     * @param price must be greater than 0
     * @return this builder
     */
    fun setPrice(price: Long): ProductBuilder {
        this.price = price
        return this
    }

    /**
     * Sets the product [partNumber], which may be null.
     *
     * @param partNumber the part number string or null
     * @return this builder
     */
    fun setPartNumber(partNumber: String?): ProductBuilder {
        this.partNumber = partNumber
        return this
    }

    /**
     * Sets the [manufactureCost], which may be null.
     *
     * @param manufactureCost the cost or null
     * @return this builder
     */
    fun setManufactureCost(manufactureCost: Long?): ProductBuilder {
        this.manufactureCost = manufactureCost
        return this
    }

    /**
     * Sets the [unitOfMeasure], which may be null.
     *
     * @param unitOfMeasure the unit or null
     * @return this builder
     */
    fun setUnitOfMeasure(unitOfMeasure: UnitOfMeasure?): ProductBuilder {
        this.unitOfMeasure = unitOfMeasure
        return this
    }

    /**
     * Sets the product [owner].
     *
     * @param owner non-null person
     * @return this builder
     */
    fun setOwner(owner: Person): ProductBuilder {
        this.owner = owner
        return this
    }

    override fun build(): Product = Product(
        name = name,
        coordinates = coordinates,
        price = price,
        partNumber = partNumber,
        manufactureCost = manufactureCost,
        unitOfMeasure = unitOfMeasure,
        owner = owner
    )
}
